### PR TITLE
fixed moving of master key encrypted and versioned files between shared folders

### LINKED ISSUE #16419: Moving files/folders from/to shared folders causes Encryption Errors
### Steps to reproduce
1. Enable Encryption
2. Let user1 share 2 folders, place some dummy content in folder1 with multiple layers of directories. (folder1/test/test2/file.txt)
3. Let user2 move the folder inside folder1 to folder2 (Using web GUI)
4. Process get stuck and Encryption errors in Logging

### Expected behaviour
Folder should move without errors

### Actual behaviour
Encryption errors

### Server configuration

**Operating system**: Debian 9

**Web server:** NGINX

**Database:** MariaDB, Redis

**PHP version:** 7.3

**Nextcloud version:** 16.0.3

**Updated from an older Nextcloud/ownCloud or fresh install:** fresh

**Where did you install Nextcloud from:** latest archive

**Signing status:** 
<details>
<summary>Signing status</summary>

```
No errors have been found.
```
</details>

**List of activated apps:**
<details>
<summary>App list</summary>

```
Enabled:
  - accessibility: 1.2.0
  - activity: 2.9.1
  - bruteforcesettings: 1.4.0
  - cloud_federation_api: 0.2.0
  - comments: 1.6.0
  - dav: 1.9.2
  - encryption: 2.4.0
  - federatedfilesharing: 1.6.0
  - federation: 1.6.0
  - files: 1.11.0
  - files_pdfviewer: 1.5.0
  - files_rightclick: 0.13.0
  - files_sharing: 1.8.0
  - files_texteditor: 2.8.0
  - files_trashbin: 1.6.0
  - files_versions: 1.9.0
  - files_videoplayer: 1.5.0
  - firstrunwizard: 2.5.0
  - gallery: 18.3.0
  - logreader: 2.1.0
  - lookup_server_connector: 1.4.0
  - nextcloud_announcements: 1.5.0
  - notifications: 2.4.1
  - oauth2: 1.4.2
  - password_policy: 1.6.0
  - privacy: 1.0.0
  - provisioning_api: 1.6.0
  - recommendations: 0.4.0
  - serverinfo: 1.6.0
  - sharebymail: 1.6.0
  - support: 1.0.0
  - survey_client: 1.4.0
  - systemtags: 1.6.0
  - theming: 1.7.0
  - twofactor_backupcodes: 1.5.0
  - twofactor_u2f: 3.0.0
  - updatenotification: 1.6.0
  - viewer: 1.0.0
  - workflowengine: 1.6.0
Disabled:
  - admin_audit
  - files_external
  - user_ldap
```
</details>

**Nextcloud configuration:**
<details>
<summary>Config report</summary>

```
{
    "system": {
        "instanceid": "***REMOVED SENSITIVE VALUE***",
        "passwordsalt": "***REMOVED SENSITIVE VALUE***",
        "secret": "***REMOVED SENSITIVE VALUE***",
        "trusted_domains": [
            "***REMOVED SENSITIVE VALUE***"
        ],
        "datadirectory": "***REMOVED SENSITIVE VALUE***",
        "dbtype": "mysql",
        "version": "16.0.3.0",
        "overwrite.cli.url": "https:\/\/nc.yisp.nl",
        "dbname": "***REMOVED SENSITIVE VALUE***",
        "dbhost": "***REMOVED SENSITIVE VALUE***",
        "dbport": "",
        "dbtableprefix": "oc_",
        "mysql.utf8mb4": true,
        "dbuser": "***REMOVED SENSITIVE VALUE***",
        "dbpassword": "***REMOVED SENSITIVE VALUE***",
        "installed": true,
        "maintenance": false,
        "memcache.locking": "\\OC\\Memcache\\Redis",
        "memcache.distributed": "\\OC\\Memcache\\Redis",
        "redis": {
            "host": "***REMOVED SENSITIVE VALUE***",
            "port": 0
        }
    }
}
```
</details>

**Are you using external storage, if yes which one:** no

**Are you using encryption:** yes

**Are you using an external user-backend, if yes which one:** no

### Client configuration
**Browser:** Chrome

**Operating system:** Ubuntu 19.04

#### Nextcloud log GUI
<details>
<summary>Nextcloud log</summary>

```
OCP\Encryption\Exceptions\GenericEncryptionException: Bad Signature
/var/www/nextcloud/apps/encryption/lib/Crypto/Crypt.php - line 467:

OCA\Encryption\Crypto\Crypt->checkSignature("d+nOgAED6pO ... E", null, "65ac461c517 ... 5")

/var/www/nextcloud/apps/encryption/lib/Crypto/Encryption.php - line 379:

OCA\Encryption\Crypto\Crypt->symmetricDecryptFileContent("*** sensiti ... *", null, "AES-256-CTR", "*** sensiti ... *", "*** sensiti ... *")

/var/www/nextcloud/lib/private/Files/Stream/Encryption.php - line 479:

OCA\Encryption\Crypto\Encryption->decrypt("*** sensiti ... *")

/var/www/nextcloud/lib/private/Files/Stream/Encryption.php - line 299:

OC\Files\Stream\Encryption->readCache()

<<closure>>

OC\Files\Stream\Encryption->stream_read(8192)

/var/www/nextcloud/3rdparty/icewind/streams/src/Wrapper.php - line 91:

fread(null, 8192)

/var/www/nextcloud/3rdparty/icewind/streams/src/CallbackWrapper.php - line 98:

Icewind\Streams\Wrapper->stream_read(8192)

<<closure>>

Icewind\Streams\CallbackWrapper->stream_read(8192)

/var/www/nextcloud/3rdparty/sabre/http/lib/Sapi.php - line 80:

stream_copy_to_stream(null, null, "12624")

/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php - line 498:

Sabre\HTTP\Sapi::sendResponse(Sabre\HTTP\Response {})

/var/www/nextcloud/3rdparty/sabre/dav/lib/DAV/Server.php - line 254:

Sabre\DAV\Server->invokeMethod(Sabre\HTTP\R ... "}, Sabre\HTTP\Response {})

/var/www/nextcloud/apps/dav/lib/Server.php - line 316:

Sabre\DAV\Server->exec()

/var/www/nextcloud/apps/dav/appinfo/v2/remote.php - line 35:

OCA\DAV\Server->exec()

/var/www/nextcloud/remote.php - line 163:

require_once("/var/www/ne ... p")
```
</details>

### PR DESCRIPTION
**Pull request by weizenspreu**
This fixes https://github.com/nextcloud/server/issues/16419 for master key encrypted files.

### CODE REVIEW

**Comment by yahesh** [comment]
This fixes the problem for master key encrypted files and improves the situation a bit for user key encrypted files.

**Comment by yahesh** [comment]
Copied over from #16419 as suggested by @kesselb:

It seems that moving **user key** encrypted folders leads into some weird locking issues in the `$this->file->getAccessList()` call of [lib/private/Files/Stream/Encryption.php:stream_open()](https://github.com/nextcloud/server/blob/11e99859ef3a28d0124241a43a934f70b7c3ff3a/lib/private/Files/Stream/Encryption.php#L264) (which we can reproduce reliably but haven't fully understood yet). At least, this locking issue doesn't break files. Copying **user key** encrypted folders works.

**This is write-down of what the general problem was:**
Let's shortly re-iterate on what happens when a file is moved from one shared folder to another shared folder:
- a copy of the file is put into the target folder
- a copy of the file is put into the trashbin of the user moving the file
- the original copy of the file is put into the trashbin of the file owner

The general logic can be found in [lib/private/Files/Storage/Wrapper/Encryption.php: moveFromStorage()](https://github.com/nextcloud/server/blob/11e99859ef3a28d0124241a43a934f70b7c3ff3a/lib/private/Files/Storage/Wrapper/Encryption.php#L606) which first copies the file/folder with the help of [lib/private/Files/Storage/Wrapper/Encryption.php:copyBetweenStorage()](https://github.com/nextcloud/server/blob/11e99859ef3a28d0124241a43a934f70b7c3ff3a/lib/private/Files/Storage/Wrapper/Encryption.php#L701) and then deletes the file/folder.

One task that has to be done is to update the `oc_filecache` table. Most of this is done by underlying classes but the `encrypted` value in the database is handled here in [lib/private/Files/Storage/Wrapper/Encryption.php:updateEncryptedVersion()](https://github.com/nextcloud/server/blob/11e99859ef3a28d0124241a43a934f70b7c3ff3a/lib/private/Files/Storage/Wrapper/Encryption.php#L661) and this is also where the problem is buried.

When `copyBetweenStorage` is first called in `moveFromStorage` it provides `true` as the `$isRename` value. This later prompts `updateEncryptedVersion` to overwrite the `encrypted` value of the source file with `1`. So back to the process:
- When the source file is copied to the target file, the `encrypted` value of the **source file** is changed into a wrong value.
- When the source file is copied to the trashbin of the user moving the file, the `encrypted` value of the source file doesn't match and the file copy fails unexpectedly.

What we did for now to remedy this problem:
- At the start of `copyBetweenStorage` we backup the original `encrypted` value.
- We let the rest of the program logic untouched because some underlying code seems to rely on this broken behaviour.
- At the end of `copyBetweenStorage` we restore the original `encrypted` value so that it is the correct value for the next copy task.

**Comment by jknockaert** [comment]
@yahesh Thanks for the PR. Ideally there would be one or more tests added to the codebase in order to avoid future regressions (and they also help to some extent in documenting the PR).
@nextcloud/encryption Reviews needed. Any other groups to invite?

**Comment by yahesh** [comment]
@jknockaert As you may have read in the corresponding issue, this problem occurs some 32 levels down into the call stack. I really have no idea how to properly mock this misbehaviour so that the requested test isn't much more than a dummy placeholder.

**Comment by jknockaert** [comment]
@yahesh Ja this is pretty low-level stuff.
@schiessle Do you have any advise as to how to design a proper test for this fix?

**Comment by papimla** [comment]
This looks like a **lot** of good work (diagnostic + solution) to solve a severe problem. It would be great to see some progress. I'm stuck with files that are most likely in a bad state because of this.

**Comment by hostingnuggets** [comment]
@papimla same here! It seems like @yahesh's work was a waste of time for him...

**Comment by kesselb** [comment]
@yahesh mind to have a look at the conflicting files? I don't know how soon one has time to review it but would be nice to have at least a recent ci run.

**Comment by papimla** [comment]
> @papimla same here! It seems like @yahesh's work was a waste of time for him...

It certainly is frustrating when good work is not put to use and is abandonned for some time. But the excellent diagnostic is still valid and will be great help to get this fixed. I would not go as far as to say it was wasted :smile:

**Comment by papimla** [comment]
@yahesh thanks a **lot** for updating the pull request !

**Comment by hostingnuggets** [comment]
@papimla exactly, now we only need to hope that the Nextclouders will take their precious time to check this PR.

**Comment by juliushaertl** [comment]
Thanks a lot @yahesh for looking into this and providing such a detailed issue description together with the fix. The changes make sense and the fix works like a charm.

**Review by juliusknorr — APPROVED** [review]

**Review by MorrisJobke — APPROVED** [review]
Code looks good 👍

**Comment by MorrisJobke** [comment]
Rebased to trigger CI again. Then this can be merged.

**Comment by MorrisJobke** [comment]
Unfortunately the tests fail :/

File: apps/files_sharing/lib/SharedStorage.php [hunk 1]
```diff
@@ -308,9 +308,9 @@ public function rename($path1, $path2) {
 		$this->init();
 		$isPartFile = pathinfo($path1, PATHINFO_EXTENSION) === 'part';
 		$targetExists = $this->file_exists($path2);
-		$sameFodler = dirname($path1) === dirname($path2);
+		$sameFolder = dirname($path1) === dirname($path2);
```
Comment (MorrisJobke) [inline apps/files_sharing/lib/SharedStorage.php:311]: Let's move this to a separate PR: #22898

File: apps/files_sharing/lib/SharedStorage.php [hunk 2]
```diff
@@ -286,7 +286,7 @@ public function fopen($path, $mode) {
 				}
 		}
 		$info = [
-			'target' => $this->getMountPoint() . $path,
+			'target' => $this->getMountPoint() . '/' . $path,
```
Comment (MorrisJobke) [inline apps/files_sharing/lib/SharedStorage.php:289]: Moved to #22902 as it is unrelated to this PR.

**Comment by MorrisJobke** [comment]
I extracted the unrelated stuff out in #22898, #22902 and #22903. After rebasing this PR and testing it again it doesn't solve the issue for me. But also on master there were already some fixes. @juliushaertl do you mind to check it out again as well? Maybe I have overseen something?

**Comment by kesselb** [comment]
@yahesh mind to do another rebase to have recent CI run?
